### PR TITLE
Improve versioning from git tags with setuptools_scm

### PR DESCRIPTION
**Pull request by rth**
One possible way to close https://github.com/icubam/icubam/issues/121 using https://github.com/pypa/setuptools_scm which determines version from git tags + commit hash.

This also adds `pyproject.toml` as a recommended way to using that package, since `setuptools_scm` now needs to be available when installing, and this ensures it. This part is not absolutely necessary, can be reverted.

The workflow is just to add tags from time to time, and the version [will be automatically generated](https://github.com/pypa/setuptools_scm#default-versioning-scheme), e.g.,
 - `0.1.8rc6` - if it's the tag exactly
 - `0.1.8rc4.dev0+g6dea7f7.d20200401` -  a few commits after the `0.1.8rc4` tag.

One limitation is that the tag needs to follow [semver](https://semver.org/), and incorrect tag scheme (for the last tag) will result in an installation error due to regex parsing such as,
```
      prefix, tail = re.match(r"(.*?)(\d+)$", version).groups()                                                                        
  AttributeError: 'NoneType' object has no attribute 'groups'
```
So for instance, `v0.1.8-deploy` is not a valid tag, while, `v0.1.8.deploy` might be OK (need to check).

Another limitation is that this package now needs to be installed before it can be used.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[build-system]
```
Comment (dulacarnold) [inline pyproject.toml:1]: Can these dependencies be simply defined in setup.py?

File: resources/config.toml [hunk 2]
```diff
@@ -1,6 +1,5 @@
 # Keys
 default_locale='fr-fr'
-version='0.1.2'
```
Comment (dulacarnold) [inline resources/config.toml:3]: wooho!

**Comment by rth** [comment]
For instance, the version currently installed in CI is `0.1.dev1+g80a4266` which should follow the following pattern (https://github.com/pypa/setuptools_scm#default-versioning-scheme),
```
{next_version}.dev{distance}+{scm letter}{revision hash}
```
the hash is different from that of the last commit because CI merged with upstream internally I think. It does work as expected for me locally.

If you don't like `setuptools_scm` and the reliance on git tags, as simpler solution could be just indicate the version in `setup.py`, and it all other places it would be retrieved from there (as done in this PR).

**Comment by dulacarnold** [comment]
@rprimet this would require us to change the deployment rule on staging.  What do you think?  This deffinitelly satisfies the 'centralize our versioning to one place' need.

**Comment by rprimet** [comment]
@dulacarnold I like it, I'll amend the versioning scheme in the deployment repository when this is merged

**Comment by rth** [comment]
> what happens if someone deploys w/out git (e.g. docker, pip install from pypi or from tgz)?

 - if you install it from PyPi, it means the the `.tar.gz` was previously generated with `python setup.py sdist`, which currently for me generates `icubam-0.1.8rc7.dev3+g7c1fe35.tar.gz` and the version is internally included in the package metadata (`icubam.egg-info/PKG-INFO`). So it will work.
 - if you install from Github with git support,
   ```
   pip install git+https://github.com/rth/icubam.git@impr-version
   ```
   e.g. to install the branch in this PR, it will also work, since you have access to git history including tags there.
- If you install from Github without git support,
  ```
  https://github.com/rth/icubam/archive/impr-version.zip
  ```
  it will fail with the following error,
  ```
  LookupError: setuptools-scm was unable to detect version for '/tmp/pip-req-build-toutfj9_'.
  
  Make sure you're either building from a fully intact git repository or PyPI tarballs. Most other sources (such as GitHub's tarballs, a git checkout without the .git folder) don't contain the necessary metadata and will not work.
  
  For example, if you're using pip, instead of https://github.com/user/proj/archive/master.zip use git+https://github.com/user/proj.git#egg=proj
  ```
  which is a reasonable and explicit error message.

So overall it should work.

File: pyproject.toml (hunk 1 above)
Comment (rth) [reply pyproject.toml:1]: We could put them in `setup.py` but it's a legacy way, that can break in some conditions. Basically `setup.py` can define dependencies fine, but cannot (very well) specify what is needed to run `setup.py` itself (i.e. build dependencies) . It's a chicken an egg problem. `pyproject.toml` was desined for that. 

See blog post in https://twitter.com/brettsky/status/1245207783240110080 for more details.

Comment (rth) [reply pyproject.toml:1]: But we can remove it, no one is enthusiastic about it.

Comment (dulacarnold) [reply pyproject.toml:1]: LGTM, let's use pyproject.toml

**Pull request merged**